// @expo/build-tools/dist/ios/fastfile.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFastfileForResigningBuild = void 0;
const path_1 = __importDefault(require("path"));
const template_file_1 = __importDefault(require("@expo/template-file"));
const RESIGN_TEMPLATE_FILE_PATH = path_1.default.join(__dirname, '../../templates/Fastfile.resign.template');
async function createFastfileForResigningBuild({ outputFile, ipaPath, signingIdentity, keychainPath, targetProvisioningProfiles, }) {
    const PROFILES = [];
    const targets = Object.keys(targetProvisioningProfiles);
    for (const target of targets) {
        const profile = targetProvisioningProfiles[target];
        PROFILES.push({
            BUNDLE_ID: profile.bundleIdentifier,
            PATH: profile.path,
        });
    }
    await (0, template_file_1.default)(RESIGN_TEMPLATE_FILE_PATH, {
        IPA_PATH: ipaPath,
        SIGNING_IDENTITY: signingIdentity,
        PROFILES,
        KEYCHAIN_PATH: keychainPath,
    }, outputFile, { mustache: false });
}
exports.createFastfileForResigningBuild = createFastfileForResigningBuild;
